### PR TITLE
[aot_autograd] Replay input mutations the way autograd saw them, and dealias marked returns

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #195352
* #195351
* __->__ #195664
* #195350
* #195349
* #195348
* #195347
* #195346

AOTAutograd's runtime wrapper had two places where the recorded metadata was
right and the runtime did something else with it.

mark_non_differentiable is keyed on TensorImpl, and the forward epilogue pairs
it with ctx._materialize_non_diff_grads = False, so marking one returned slot
marks every slot holding that same object, and each is then handed None
rather than zeros. Backends share one object across slots freely -- inductor
lowers aten.detach to a no-op, and h * 1 folds away -- so `return y[0:4],
y[4:8], y.detach()` marks the intermediate base the backward needs a tangent
for, and `return h * 1.0, h.detach()` marks the only differentiable output.
The recorded metadata is not at fault, the runtime identity collision is, so a
slot about to be marked that shares its object with a slot staying
differentiable is detached first and gets its own TensorImpl; one that collides
with nothing keeps its identity. Sparing the aliasing slot instead, by
declining to mark it, would leave the detach output with a grad_fn. The marked
index set is bound as a codegen-time constant so the per-call cost is one
identity scan over the Tensor slots.

The mutation epilogue wrote every functionalized input mutation back as a
tracked copy_. That is right for a mutation autograd saw, and wrong for one it
did not. Functionalization records a write as mutations_hidden_from_autograd
only when the op calls mark_mutation_hidden_from_autograd, which today is the
triton kernel HOP alone: such a write bypasses the ADInplaceOrView layer that
decides whether an eager in-place op bumps the version counter, so it left no
trace on the tensor in eager, and replaying it as a tracked copy_ both
rewrites the tensor's grad_fn and, when the target is an input a custom
autograd.Function returned as-is (an identity view stamped
IN_CUSTOM_FUNCTION), raises where eager did not. The epilogue now follows the
recorded metadata, exactly as apply_in_graph_mutations already does for
in-graph mutations: a hidden mutation is written under no_grad with the
version counter preserved, a mutation recorded under no_grad or inference
mode is written under no_grad, and everything else stays a tracked copy_.
That split now applies to leaf inputs too; previously a requires-grad leaf
always took the detach().copy_() path, which bumps a version counter a hidden
write never bumped. A tracked requires-grad leaf keeps detach().copy_(): a tracked write reaches
such a leaf through a detached view or through .data, and both take that
path, so a .data write that eager never counted still bumps the version once
under compile, exactly as it did before this change.
A Tensor(a!) op registered through torch.library with a
CompositeExplicitAutograd impl and no ADInplaceOrView kernel, writing through
.data, is recorded as a tracked mutation, so compiled code raises on a
restricted view where eager does not; that is what main did before this
stack, and teaching functionalization to record such ops as hidden is a
separate change. Deciding from the target's CreationMeta instead was rejected
because nothing guards it and it would let a tracked mutation silently
succeed on a view eager refuses. The reference _apply_input_mutations
implementation follows the same rule as the codegen'd epilogue. Both also check
for a mutated input bound to a stream index before the leaf branch rather than
after it, so that write is refused wherever it appears, and a mutation of an
inference tensor is replayed without the version-counter guard, which inference
tensors do not carry.

Diagnostic only: a kept tangent slot arriving as None now raises from
process_runtime_tangent, naming the forward output behind the slot, which of
the returned slots it is, and its recorded dtype, rather than surfacing several
frames lower as inductor's "expected Tensor()" against a codegen'd name like
tangents_3, which is not positional and cannot be mapped back to a user output
by hand. That needs the tangent dtypes, which make_runtime_safe dropped along
with the fake tangents, so ViewAndMutationMeta records them; a length mismatch
against the kept slots is logged rather than silently treated as unknown.

These changes run for every torch.compile and are not gated on precompile,
which is why they are their own commit rather than part of the precompile
stack's own PRs; nothing later in the stack calls into them directly.

Test Plan:

```
python test/functorch/test_aotdispatch.py
python test/functorch/test_codegen_mutation_epilogue.py
```

The two aliasing tests compare requires_grad, grad_fn and the gradient against
uncompiled eager over aot_eager and inductor, and the intermediate-base one
also asserts the detach output stays a leaf so the rejected fix cannot pass
it. The mutation tests split three ways. End to end,
test_tracked_input_mutation_on_custom_function_view_raises warms the compiled
function on an ordinary tensor and then calls it with the restricted view,
asserting the tracked write raises exactly as eager does on aot_eager and
inductor. test_hidden_input_mutation_replayed_onto_custom_function_view calls
the _replay_input_mutation helper directly on such a view and asserts the
grad_fn, the version counter and the custom backward all survive. The codegen
tests in test_codegen_mutation_epilogue.py assert the emitted epilogue is
copy_, a no_grad copy_, detach().copy_() or _replay_input_mutation according
to the recorded metadata, including a hidden and a no_grad mutation of a
requires-grad leaf, with the hidden flag patched in since only the triton
kernel HOP sets it. Three more cover the None-tangent path, and a warm-cache
test asserts the recorded tangent dtypes survive an AOTAutogradCache reload.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this content lives in #195357, #195358 and #195359 (the raise on a custom-Function view became warn-and-replay, deliberately). Closing.
